CylBinaryCompactObject: add equiangular map.

markscheel opened the pull request:
## Proposed changes

Adds an equiangular map option to CylindricalBinaryCompactObject.

### Upgrade instructions

<!-- UPGRADE INSTRUCTIONS -->
When using the `CylindricalBinaryCompactObject` domain creator one must now specify `UseEquiangularMap`.  Setting this option to false gives the previous behavior.
<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

markscheel committed:
CylBinaryCompactObject: add equiangular map.

nilsvu commented on src/Domain/Creators/CylindricalBinaryCompactObject.cpp:
```diff
@@ -63,6 +63,7 @@ CylindricalBinaryCompactObject::CylindricalBinaryCompactObject(
     typename IncludeInnerSphereB::type include_inner_sphere_B,
     typename IncludeOuterSphere::type include_outer_sphere,
     typename OuterRadius::type outer_radius,
+    typename UseEquiangularMap::type use_equiangular_map,
```
Just write `bool`, no need to hide the type. We've removed this unnecessary obscuration in the other domain creators too (except for complicated `std::variant` types). A separate commit is fine.

nilsvu commented on tests/Unit/Domain/Creators/Test_CylindricalBinaryCompactObject.cpp:
```diff
@@ -364,8 +368,11 @@ void test_connectivity() {
           CAPTURE(include_inner_sphere_B);
-          test_connectivity_once(with_sphere_e, include_inner_sphere_A,
-                                 include_inner_sphere_B, include_outer_sphere);
+          for (const bool use_equiangular_map : {true, false}) {
```
Use the new `cartesian_product` you avoid the nested loops. Also CAPTURE the bool.

markscheel replied:
I capture the bool now.

I don't see how to use `cartesian_product` except for doing things that would make the code unreadable and/or bug prone, like creating a `cartesian_product({{false}}, {{true, false}}, {{true, false}}, {{true, false}},{{true, false}})` and then looping over it and trying to name the different fields `include_inner_sphere_A` etc (and get the order right) and CAPTURE them.
Maybe `cartesian_product` needs better documentation or an example?

nilsvu reviewed (changes requested):
Are you relatively certain that the unit tests would catch bugs where coord maps join discontinuously because of missing equiangular maps? For example, do they fail when you only make some of the cylinders equiangular?

markscheel committed:
CylBinaryCO: Basic types in ctor, members.

Previously was using `typename CenterA::type center_A`, etc.

markscheel commented:
I tried making only some of the wedges equiangular, or some of the cylinders equiangular, and the test still passes. I don't understand why `TestHelpers::domain::creators::test_domain_creator` doesn't test this.  What happens for the other BBH domain? Is that untested as well?

Made other changes except for cartesian_product.

kidder commented:
> I tried making only some of the wedges equiangular, or some of the cylinders equiangular, and the test still passes. I don't understand why `TestHelpers::domain::creators::test_domain_creator` doesn't test this. What happens for the other BBH domain? Is that untested as well?
> 
> Made other changes except for cartesian_product.

try cherry-picking the commit from https://github.com/kidder/spectre/tree/check_block_faces and see if the test still passes

markscheel commented:
Rebased, which also did some of the requested changes here. So now there are only 3 relevant fixup commits.
I will try https://github.com/kidder/spectre/tree/check_block_faces

markscheel commented:
Thanks @kidder !  when I cherry-pick  https://github.com/kidder/spectre/tree/check_block_faces, the current PR passes the tests as-is, but if I intentionally introduce a bug by turning on `use_equiangular_map` for some blocks but not others, then the test fails.

kidder commented:
okay, I'll clean that up and make a PR

nilsvu reviewed (dismissed):
I thought that continuity of maps was tested in one of the generic tests already, but apparently it isn't. Thanks for checking. We don't need to hold up the PR with that. @kidder thanks for adding that test.

kidder commented:
the current test only checks the corners of the block faces, not the interior...

markscheel commented:
squashed

nilsvu reviewed (changes requested):
The test timed out in one of the builds. I've seen timeouts in this test before. Can we reduce the amount of work done in the big cartesian product of options?
Edit: or we randomly skip some of the iterations of the loop ("fuzzy cartesian product")

nilsvu commented:
@markscheel would you mind adding the random-skipping code also to the BinaryCompactObject domain test while you rebase/squash? It's been timing out as well.

markscheel committed:
Test_BinaryCompObj: ignore some tests randomly

The idea is to speed up the test by not testing
every possible combination of the options.
